### PR TITLE
Use NFI for setting the process title on Darwin

### PR DESCRIPTION
**Pull request by chrisseaton**

### CODE REVIEW

File: src/main/ruby/core/process.rb [hunk 1]
```diff
@@ -124,11 +124,22 @@ def self.cpu_times
...
+    elsif Rubinius.linux? && File.readable?('/proc/self/maps')
+      setproctitle_linux_from_proc_maps(title)
+    elsif Rubinius.darwin?
+      # When we call _NSGetArgv we seem to always get a string that looks like what we'd expect from running ps, but
+      # with a null character inserted early. I don't know where this comes from, but it means I don't know how to get
+      # the length of space available for writing in the new program name. We therefore limit to about 40 characters,
+      # which is a number without any foundation, but it at least allows the specs to pass, the functionality to be
+      # useful, and probably avoid crashing anyone's programs. I can't pretend this is great engineering.
+      name_address = Truffle::POSIX._NSGetArgv.read_pointer.read_pointer
+      name_address.write_string_length title.encode('utf-8'), 40
```
Comment (eregon) [inline src/main/ruby/core/process.rb]: You should not need to encode title here, just passing as-is should be enough.

File: src/main/ruby/core/process.rb [hunk 2]
```diff
@@ -124,11 +124,22 @@ def self.cpu_times
...
-      setproctitle_linux_from_proc_maps(val)
+    title = Rubinius::Type.coerce_to(title, String, :to_str)
+    if Truffle.aot?
+      Truffle.invoke_primitive(:aot_set_process_title, title)
+    elsif Rubinius.linux? && File.readable?('/proc/self/maps')
+      setproctitle_linux_from_proc_maps(title)
+    elsif Rubinius.darwin?
+      # When we call _NSGetArgv we seem to always get a string that looks like what we'd expect from running ps, but
+      # with a null character inserted early. I don't know where this comes from, but it means I don't know how to get
+      # the length of space available for writing in the new program name. We therefore limit to about 40 characters,
```
Comment (eregon) [inline src/main/ruby/core/process.rb]: Yes, in native memory argv are simply separated by the strings null bytes.
One could read the string at the current pointer or call `strlen()` to know how long `argv[0]` is.
This can be another PR of course.

Comment (eregon) [reply src/main/ruby/core/process.rb]: You would need to also know `argc` to know how long is the full command line I think.

File: src/main/ruby/core/process.rb (hunk 1 above)
Comment (chrisseaton) [reply src/main/ruby/core/process.rb]: What if it's in some crazy encoding though?

**Comment by chrisseaton** [comment]
@eregon review again please.

File: src/main/ruby/core/process.rb (hunk 1 above)
Comment (eregon) [reply src/main/ruby/core/process.rb]: MRI doesn't care, it allows to set it to a UTF-16BE String.

Comment (eregon) [reply src/main/ruby/core/process.rb]: Please remove the `encode` call, it's incompatible with MRI and the Linux logic below.

Comment (chrisseaton) [reply src/main/ruby/core/process.rb]: Done.
